### PR TITLE
[webui] Scalability fixes for the task timeline and visualizations

### PR DESCRIPTION
**Pull request by ericl**
This PR ensures web ui responsiveness (https://github.com/ray-project/ray/issues/933) by adding a hard cap on the number of tasks returned (10000).

You can test with the following snippet which launches one million tasks locally. The UI should respond fluidly in all the visualizations except the timeline, which still should update within about 10 seconds.
```
import time
import ray

@ray.remote
def f(x):
    return x

print("Launching tasks")

ray.init()
for x in range(1000000):
    y = f.remote(x)
ray.get(y)
print("Done")

time.sleep(99999999)
```

cc @alanamarzoev @robertnishihara

### CODE REVIEW

File: python/ray/experimental/state.py [hunk 1]
```diff
@@ -562,8 +571,10 @@ def micros(ts):
         def micros_rel(ts):
             return micros(ts - start_time)
 
-        task_profiles = self.task_profiles(start=0, end=time.time())
```
Comment (ericl) [inline python/ray/experimental/state.py]: Note that this was performing a full table scan.

File: python/ray/experimental/state.py [hunk 2]
```diff
@@ -562,8 +571,10 @@ def micros(ts):
...
-        task_table = self.task_table()
```
Comment (ericl) [inline python/ray/experimental/state.py]: This was also a full table scan.

File: python/ray/experimental/state.py [hunk 3]
```diff
@@ -649,14 +660,16 @@ def micros_rel(ts):
...
                     parent = {
                         "cat": "submit_task",
                         "pid": "Node " + parent_worker["node_ip_address"],
                         "tid": parent_info["worker_id"],
-                        "ts": micros_rel(task_profiles[task_table[task_id]
-                                                       ["TaskSpec"]
-                                                       ["ParentTaskID"]]
-                                          ["get_arguments_start"]),
+                        "ts": micros_rel(
+                            parent_profile and
```
Comment (ericl) [inline python/ray/experimental/state.py]: With truncation it's possible we don't have some parent profiles. We are just ignoring them for now.

File: python/ray/experimental/state.py [hunk 4]
```diff
@@ -419,17 +422,25 @@ def task_profiles(self, start=None, end=None, num_tasks=None, fwd=True):
         task_info = dict()
         event_log_sets = self.redis_client.keys("event_log*")
 
+        if num_tasks is None:
```
Comment (pcmoritz) [inline python/ray/experimental/state.py]: This code should not be part of the GlobalState API and rather num_tasks should be passed in from the webui

File: python/ray/experimental/state.py (hunk 2 above)
Comment (pcmoritz) [reply python/ray/experimental/state.py]: There is actually a way to get the best of both worlds with the SCAN [1] class of functions from redis (I guess that's what you mean with the TODO above, do you want to make it a little more precise?)

[1] https://redis.io/commands/scan

File: python/ray/experimental/state.py (hunk 4 above)
Comment (ericl) [reply python/ray/experimental/state.py]: Done

File: python/ray/experimental/state.py (hunk 2 above)
Comment (ericl) [reply python/ray/experimental/state.py]: SCAN won't do it, but MGET might. I think the refactoring is a bit nontrivial but I updated the TODO.

Comment (pcmoritz) [reply python/ray/experimental/state.py]: You are right, sounds good!

**Comment by ericl** [comment]
Looks like these are broken tests, I'll take a look tomorrow.

**Comment by ericl** [comment]
@pcmoritz looks like one tests are passing. One of the travis builds was cancelled but I don't see failures.

**Comment by robertnishihara** [comment]
Nice! I'm trying it out now.

**Comment by robertnishihara** [comment]
I think I can generate key errors when I use timelines with task dependencies (the line numbers are sligtly off because I rebased the PR locally).

```python
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
~/Workspace/ray/python/ray/experimental/ui.py in handle_submit(sender)
    442                                              breakdowns=breakdown,
    443                                              obj_dep=obj_dep.value,
--> 444                                              task_dep=task_dep.value)
    445         print("Opening html file in browser...")
    446 

~/Workspace/ray/python/ray/experimental/state.py in dump_catapult_trace(self, path, task_info, breakdowns, task_dep, obj_dep)
    748                             owner_task = self._object_table(arg)["TaskID"]
    749                             owner_worker = (workers[
--> 750                                 task_info[owner_task]["worker_id"]])
    751                             # Adding/subtracting 2 to the time associated with
    752                             # the beginning/ending of the flow event is

KeyError: 'be338b34b87c7497d6fd82768b5f3e3d5f4442c3'
```

E.g.,

```python
import ray
ray.init()

@ray.remote
def f(x):
    return 1

x = 1
for _ in range(10 ** 5):
    x = f.remote(x)
```

Then visualize the last 10000 tasks.

Presumably this arises from omitting the initial scan of the entire task table.

**Comment by ericl** [comment]
Alright that should be fixed.

**Comment by robertnishihara** [comment]
I don't see any new commits, did you push something?

**Comment by ericl** [comment]
Pushed

**Comment by robertnishihara** [comment]
Looks good to me, I'll merge it once the tests run.

**Pull request merged**